drivers/vl53l1x/vl53l1x.c

```diff
@@ -0,0 +1,1292 @@
+
+    /*
+     * from VL53L1_preset_mode_timed_ranging_*
+     * GPH is 0 after reset, but writing GPH0 and GPH1 above seem to set GPH
+     * to 1, and things don't seem to work if we don't set GPH back to 0 (which
+     * the API does here).
+     */
+    EXEC_RET(_write_byte(dev, VL53L1X_SYSTEM__GROUPED_PARAMETER_HOLD, 0x00));
+    EXEC_RET(_write_byte(dev, VL53L1X_SYSTEM__SEED_CONFIG, 1));
+
+    /* from VL53L1_config_low_power_auto_mode */
+    EXEC_RET(_write_byte(dev, VL53L1X_SYSTEM__SEQUENCE_CONFIG, 0x8B));
+    EXEC_RET(_write_word(dev, VL53L1X_DSS_CONFIG__MANUAL_EFFECTIVE_SPADS_SELECT, 200 << 8));
+    EXEC_RET(_write_byte(dev, VL53L1X_DSS_CONFIG__ROI_MODE_CONTROL, 2));
+
+    /* default to long range, 50 ms timing budget */
+    EXEC_RET(vl53l1x_set_distance_mode(dev, VL53L1X_DIST_LONG));
+    EXEC_RET(vl53l1x_set_timing_budget(dev, 50000));
+
+    /*
+     * the API triggers this change in VL53L1_init_and_start_range() once a
+     * measurement is started; assumes MM1 and MM2 are disabled
+     */
+    uint16_t word;
+    EXEC_RET(_read_word(dev, VL53L1X_MM_CONFIG__OUTER_OFFSET_MM, &word));
+    word *= 4;
+    EXEC_RET(_write_word(dev, VL53L1X_ALGO__PART_TO_PART_RANGE_OFFSET_MM, word));
+
+    /* Copyrighted POLOLU / ST API code end */
+
+    #endif /* MODULE_VL53L1X_ST_API */
+
+    /* set measurement parameters */
+    EXEC_RET(vl53l1x_set_distance_mode(dev, dev->params.mode));
+    EXEC_RET(vl53l1x_set_timing_budget(dev, dev->params.budget));
+    EXEC_RET(vl53l1x_set_measurement_period(dev, dev->params.period));
+
+    #endif /* MODULE_VL53L1X_BASIC */
+
+    /* start measurement */
+    EXEC_RET(_start(dev));
+
+    return VL53L1X_OK;
+}
+
+#define VL53L1X_BOOT_TIMEOUT    10      /* 10 ms */
+
+static int _reset(const vl53l1x_t *dev)
+{
+    #if MODULE_VL53L1X_ST_API
+
+    /* write reset impuls for 150 us */
+    VL53L1_software_reset((VL53L1_Dev_t*)dev);
+
+    /* give the sensor some time to boot */
+    xtimer_usleep(4 * US_PER_MS);
+
+    /* wait until the sensor has been booted */
+    EXEC_RET(VL53L1_WaitDeviceBooted((VL53L1_Dev_t*)dev));
+
+    /* set VDDIO */
+    EXEC_RET(VL53L1_UpdateByte((VL53L1_Dev_t*)dev,
+                               VL53L1X_PAD_I2C_HV__EXTSUP_CONFIG,
+                               0xfe, dev->params.vddio_2v8 ? 1 : 0));
+
+    #else /* MODULE_VL53L1X_ST_API */
+
+    /* write reset impuls of 100 us */
+    _write_byte(dev, VL53L1X_SOFT_RESET, 0);
+    xtimer_usleep(100);
+    _write_byte(dev, VL53L1X_SOFT_RESET, 1);
+
+    /* wait until the sensor has been booted in 1 ms steps*/
+    unsigned timeout = VL53L1X_BOOT_TIMEOUT;
+    uint8_t res;
+    while (timeout--) {
+        if ((_read_byte(dev, VL53L1X_FIRMWARE__SYSTEM_STATUS, &res) == VL53L1X_OK) &&
+            (res & 0x01) == 1) {
+            break;
+        }
+        xtimer_usleep(1 * US_PER_MS);
+    }
+    if (timeout == 0) {
+        DEBUG_DEV("could not boot the sensor", dev);
+        return VL53L1X_ERROR_TIMEOUT;
+    }
```

Comment:
Are you sure that the postdecrement in the `while()` doesn't trigger after `timeout` has reached zero? I have the gutfeeling that it does.